src/stan/mcmc/hmc/xhmc/base_xhmc.hpp

```diff
@@ -0,0 +1,223 @@
+#ifndef STAN_MCMC_HMC_NUTS_BASE_XHMC_HPP
+#define STAN_MCMC_HMC_NUTS_BASE_XHMC_HPP
+
+#include <stan/interface_callbacks/writer/base_writer.hpp>
+#include <boost/math/special_functions/fpclassify.hpp>
+#include <stan/mcmc/hmc/base_hmc.hpp>
+#include <stan/mcmc/hmc/hamiltonians/ps_point.hpp>
+#include <algorithm>
+#include <cmath>
+#include <limits>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace mcmc {
+    /**
+      * Exhaustive Hamiltonian Monte Carlo (XHMC) with multinomial sampling
+    */
+    template <class Model, template<class, class> class Hamiltonian,
+              template<class> class Integrator, class BaseRNG>
+    class base_xhmc : public base_hmc<Model, Hamiltonian, Integrator, BaseRNG> {
+    public:
+      base_xhmc(Model &model, BaseRNG& rng)
+        : base_hmc<Model, Hamiltonian, Integrator, BaseRNG>(model, rng),
+          depth_(0), max_depth_(5), max_deltaH_(1000), x_delta_(0.1),
+          n_leapfrog_(0), divergent_(0), energy_(0) {
+      }
+
+      ~base_xhmc() {}
+
+      void set_max_depth(int d) {
+        if (d > 0)
```

Comment:
I really think these should either throw exceptions or be eliminated and the checking left up to the caller.